### PR TITLE
Allow parallel tasks to exit on any completion

### PR DESCRIPTION
**Pull request by b-deam**
With this commit we add the ability to specify a new `any` value
for the `completed-by` param for `parallel` task blocks. This 
allows the `parallel` block to be completed by whichever task is
first to complete successfully.

Closes https://github.com/elastic/rally/issues/1261

Tested with track:
```
{
  "description": "'any' testing ",
  "challenges": [
    {
      "name": "parallel-any",
      "description": "Track completed-by property",
      "schedule": [
        {
          "name": "outside-cluster-health",
          "operation": {
            "operation-type": "cluster-health"
          }
        },
        {
          "parallel": {
            "completed-by": "any",
            "tasks": [
              {
                "name": "check-cluster-health",
                "operation": {
                  "operation-type": "cluster-health"
                }
              },
              {
                "name": "run-node-stats",
                "operation": {
                  "operation-type": "node-stats"
                },
                "iterations": 300000000
              }
            ]
          }
        }
      ]
    }
  ]
}
```

### CODE REVIEW

File: docs/track.rst [hunk 1]
```diff
@@ -604,6 +604,49 @@ All tasks in the ``schedule`` list are executed sequentially in the order in whi
...
+              "parallel": {
+                "completed-by": "any",
+                "tasks": [
+                  {
+                    "name": "bulk-task-1",
+                    "operation": {
+                      "operation-type": "bulk",
+                      "bulk-size": 1000
+                    },
+                  "clients": 8
```
Comment (DJRickyB) [inline docs/track.rst:633]: nit - indentation
```suggestion
                    "clients": 8
```

File: docs/track.rst [hunk 2]
```diff
@@ -604,6 +604,49 @@ All tasks in the ``schedule`` list are executed sequentially in the order in whi
...
 
+.. warning::
+
+    If the value ``any`` is provided for the ``completed-by`` parameter, the first completion of a client task in the ``parallel`` block will cause all remaining clients and tasks within the ``parallel`` block to immediately exit without waiting for completion.
+
+    For example, given the below track:
+
+        1. Both ``bulk-task-1`` and ``bulk-task-2`` execute in parallel
+        2. ``bulk-task-1`` Client 1/8 is first to complete its assigned partition of work
+        3. ``bulk-task-1`` will now cause the ``parallel`` task to complete and **not** wait for either the remaining 7/8 ``bulk-task-1``'s clients to complete, or for any of ``bulk-task-2``'s clients to complete
```
Comment (DJRickyB) [inline docs/track.rst:615]: isn't this contrary to
> A task is completed if and only if all associated clients have completed execution.

?

File: esrally/track/loader.py [hunk 3]
```diff
@@ -1497,6 +1496,8 @@ def parse_parallel(self, ops_spec, ops, challenge_name):
                         f"'parallel' element for challenge '{challenge_name}' contains multiple tasks with the "
                         f"name '{completed_by}' marked with 'completed-by' but only task is allowed to match."
                     )
+                elif task.any_completes_parent:
+                    completion_task = "any"
```
Comment (danielmitterdorfer) [inline esrally/track/loader.py:1500]: We use `completion_task` only as a flag whether there is a completion task so a specific name is actually not required. Can we refactor this to only set a flag instead in both cases (any task completes the parallel structure and specific task completes the parallel structure)?

**Review by danielmitterdorfer — COMMENTED** [review]
Thanks for the PR! I left a few suggestions. IMHO also a test for the track loader is missing.

**Comment by b-deam** [comment]
I _think_ I addressed the comments, can you take a look when you get a chance?

File: esrally/track/loader.py [hunk 4]
```diff
@@ -1487,17 +1487,17 @@ def parse_parallel(self, ops_spec, ops, challenge_name):
                     )
 
         if completed_by:
-            completion_task = None
+            completion_task = False
```
Comment (danielmitterdorfer) [inline esrally/track/loader.py:1490]: Nit: Based on the variable name it's not clear that this is a flag. How about `has_completion_task`?

**Review by danielmitterdorfer — APPROVED** [review]
Thanks for iterating. I left one nit but other than that LGTM. No need for another review round.

**Pull request merged**